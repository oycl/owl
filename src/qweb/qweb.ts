import { CompiledTemplate, compileTemplate } from "./compiler";

// -----------------------------------------------------------------------------
// Global template Map
// -----------------------------------------------------------------------------

let nextId = 1;
const templateMap: { [name: string]: string } = {};

export function addTemplate(name: string, template: string) {
  templateMap[name] = template;
}

export function xml(strings, ...args) {
  const name = `__template__${nextId++}`;
  const value = String.raw(strings, ...args);
  addTemplate(name, value);
  return name;
}

// -----------------------------------------------------------------------------
// QWeb
// -----------------------------------------------------------------------------

export const utils: any = {};

export const compiledTemplates: { [name: string]: CompiledTemplate } = {};

export function getTemplateFn(template: string): CompiledTemplate {
  let fn = compiledTemplates[template];
  if (!fn) {
    const rawTemplate = templateMap[template];
    if (rawTemplate === undefined) {
      throw new Error("qweb not implemented yet...");
    }

    fn = compileTemplate(template, rawTemplate);
    compiledTemplates[template] = fn;
  }
  return fn;
}
