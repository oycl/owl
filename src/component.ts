import { Fiber } from "./fiber";
import { getTemplateFn, utils as qwebUtils } from "./qweb/qweb";
import { CompiledTemplate, RenderContext } from "./qweb/compiler";
import { scheduler } from "./scheduler";
import { patch, VDataNode, NodeType } from "./vdom";
import { escape } from "./utils";

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

export interface FunctionComponent {
  template: string;
  name?: string;
  setup?: (props: any, env: any) => any | void | Promise<any | void>;
}

export interface ComponentData {
  fiber: Fiber;
  templateFn: CompiledTemplate;
  context: any;
}

export interface FnInstance {
  vtree: VTree;
  context: any;
}

export type VTree = VDataNode<ComponentData>;

// -----------------------------------------------------------------------------
// mount
// -----------------------------------------------------------------------------

export class Component<Props = any, Env = any> {
  static template: string;
  props: Props;
  env: Env;

  el: HTMLElement | Text | Comment | null = null;
  __vtree: VTree | null = null;

  constructor(props: Props, env: Env) {
    this.props = props;
    this.env = env;
  }
}

type MountTarget = HTMLElement | DocumentFragment;

interface MountOptions {
  props?: Object;
  env?: Object;
}

export function mount(
  target: MountTarget,
  fn: FunctionComponent,
  options?: MountOptions
): Promise<FnInstance>;
export function mount<C extends typeof Component>(
  target: MountTarget,
  Comp: C,
  options?: MountOptions
): Promise<InstanceType<C>>;
export function mount(
  target: MountTarget,
  comp: Component,
  options?: MountOptions
): Promise<Component>;
export function mount(
  target: MountTarget,
  fn: FnInstance,
  options?: MountOptions
): Promise<FnInstance>;
export async function mount(
  target: MountTarget,
  elem: any,
  options: MountOptions = {}
): Promise<any> {
  let tree: VTree;
  let result: any = null;
  if (!(target instanceof HTMLElement || target instanceof DocumentFragment)) {
    const name = elem instanceof Component ? elem.constructor.name : elem.name || "Undefined";
    let message = `Component '${name}' cannot be mounted: the target is not a valid DOM node.`;
    message += `\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)`;
    throw new Error(message);
  }
  if (elem instanceof Component) {
    return scheduler.addFiber(elem.__vtree!.data.fiber).then(() => {
      target.appendChild(elem.el!);
      return elem;
    });
  }
  if (elem.prototype instanceof Component || elem === Component) {
    tree = makeClassComponent(elem, options);
  } else if (elem.vtree) {
    tree = elem.vtree;
  } else {
    tree = makeFnComponent(elem, options);
    result = {
      vtree: tree,
      context: tree.data.context,
    };
  }
  return scheduler.addFiber(tree.data.fiber).then(() => {
    const fragment = document.createDocumentFragment();
    patch(fragment, tree);
    target.appendChild(fragment);
    return result || tree.data.context;
  });
}

function makeFnComponent(fn: FunctionComponent, options: MountOptions): VTree {
  let template: string = fn.template;
  if (!template) {
    const name = fn.name || "Anonymous Function Component";
    throw new Error(`Component "${name}" does not have a template defined!`);
  }
  const fiber = new Fiber(null);
  fiber.counter++;
  const props = options.props || {};
  const env = options.env || {};
  const context = fn.setup ? fn.setup(props, env) : {};
  const data: ComponentData = {
    fiber,
    context,
    templateFn: getTemplateFn(fn.template),
  };
  const tree: VTree = {
    type: NodeType.Data,
    data,
    child: null,
    key: 1,
    hooks: {},
  };
  new Promise(async (resolve) => {
    tree.data.templateFn.call(qwebUtils, tree, context);
    fiber.counter--;
    resolve();
  });
  return tree;
}

function makeClassComponent(C: typeof Component, options: MountOptions): VTree {
  let template: string = C.template;
  if (!template) {
    throw new Error(`Component "${C.name}" does not have a template defined!`);
  }
  const props = options.props || {};
  const env = options.env || {};
  const c = new C(props, env);
  const fiber = new Fiber(null);
  fiber.counter++;
  const data: ComponentData = {
    fiber,
    context: c,
    templateFn: getTemplateFn(template),
  };
  const tree: VTree = {
    type: NodeType.Data,
    data,
    child: null,
    key: 1,
    hooks: {},
  };
  c.__vtree = tree;
  tree.hooks.create = (el) => (c.el = el);
  new Promise((resolve) => {
    tree.data.templateFn.call(qwebUtils, tree, c);
    fiber.counter--;
    resolve();
  });
  return tree;
}

qwebUtils.makeComponent = function (parent: VTree, name: string, context: RenderContext): VTree {
  const definition = context[name];
  if (definition instanceof Component) {
    throw new Error("not done yet");
  } else {
    return makeFnComponent(definition, {});
  }
};

/**
 * Render a template to a html string.
 *
 * Note that this is more limited than the `render` method: it is not suitable
 * to render a full component tree, since this is an asynchronous operation.
 * This method can only render templates without components.
 */
export function renderToString(name: string, context: RenderContext = {}): string {
  const fn = getTemplateFn(name);
  const tree: VTree = {
    type: NodeType.Data,
    data: {} as any,
    child: null,
    key: 1,
    hooks: {},
  };
  fn.call(qwebUtils, tree, context);
  const div = document.createElement("div");
  patch(div, tree);

  function escapeTextNodes(node) {
    if (node.nodeType === 3) {
      node.textContent = escape(node.textContent);
    }
    for (let n of node.childNodes) {
      escapeTextNodes(n);
    }
  }
  escapeTextNodes(div);
  return div.innerHTML;
}
