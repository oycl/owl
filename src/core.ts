import { Component } from "./component";
import { Fiber } from "./fiber";
import { getTemplateFn, utils as qwebUtils } from "./qweb/qweb";
import { CompiledTemplate, RenderContext } from "./qweb/compiler";
import { scheduler } from "./scheduler";
import { patch, VDataNode, NodeType } from "./vdom";
import { escape } from "./utils";

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

interface FunctionComponent {
  template: string;
  setup?: (props: any) => any | void | Promise<any | void>;
}

export interface ComponentData {
  fiber: Fiber;
  templateFn: CompiledTemplate;
  context: any;
}

export type VTree = VDataNode<ComponentData>;

// -----------------------------------------------------------------------------
// mount
// -----------------------------------------------------------------------------

type MountTarget = HTMLElement;

export function mount<Ctx>(target: MountTarget, fn: FunctionComponent): Promise<Ctx>;
export function mount<C extends typeof Component>(
  target: MountTarget,
  Comp: C
): Promise<InstanceType<C>>;
export async function mount(target: MountTarget, elem: any): Promise<any> {
  let tree: VTree;

  if (elem.prototype instanceof Component) {
    tree = makeClassComponent(elem);
  } else {
    tree = makeFnComponent(elem);
  }
  return scheduler.addFiber(tree.data.fiber).then(() => {
    const fragment = document.createDocumentFragment();
    patch(fragment, tree);
    target.appendChild(fragment);
    return tree.data.context;
  });
}

function makeFnComponent(fn: FunctionComponent): VTree {
  const fiber = new Fiber(null);
  fiber.counter++;
  const context = fn.setup ? fn.setup({}) : {};
  const data: ComponentData = {
    fiber,
    context,
    templateFn: getTemplateFn(fn.template),
  };
  const tree: VTree = {
    type: NodeType.Data,
    data,
    child: null,
    key: 1,
    hooks: {},
  };
  new Promise(async (resolve) => {
    tree.data.templateFn.call(qwebUtils, tree, context);
    fiber.counter--;
    resolve();
  });
  return tree;
}
function makeClassComponent(C: typeof Component): VTree {
  let template: string = C.template;
  const c = new C();
  const fiber = new Fiber(null);
  fiber.counter++;
  const data: ComponentData = {
    fiber,
    context: c,
    templateFn: getTemplateFn(template),
  };
  const tree: VTree = {
    type: NodeType.Data,
    data,
    child: null,
    key: 1,
    hooks: {},
  };
  tree.hooks.create = (el) => (c.el = el);
  new Promise((resolve) => {
    tree.data.templateFn.call(qwebUtils, tree, c);
    fiber.counter--;
    resolve();
  });
  return tree;
}

qwebUtils.makeComponent = function (parent: VTree, name: string, context: RenderContext): VTree {
  const definition = context[name];
  if (definition instanceof Component) {
    throw new Error("not done yet");
  } else {
    return makeFnComponent(definition);
  }
};

/**
 * Render a template to a html string.
 *
 * Note that this is more limited than the `render` method: it is not suitable
 * to render a full component tree, since this is an asynchronous operation.
 * This method can only render templates without components.
 */
export function renderToString(name: string, context: RenderContext = {}): string {
  const fn = getTemplateFn(name);
  const tree: VTree = {
    type: NodeType.Data,
    data: {} as any,
    child: null,
    key: 1,
    hooks: {},
  };
  fn.call(qwebUtils, tree, context);
  const div = document.createElement("div");
  patch(div, tree);

  function escapeTextNodes(node) {
    if (node.nodeType === 3) {
      node.textContent = escape(node.textContent);
    }
    for (let n of node.childNodes) {
      escapeTextNodes(n);
    }
  }
  escapeTextNodes(div);
  return div.innerHTML;
}
