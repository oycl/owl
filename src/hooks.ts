import { reactive } from "./reactive";
import { engine } from "./core/rendering_engine";

export function useState<T>(state: T): T {
  let root = engine.currentVTree!;
  return reactive(state, () => engine.render(root));
}
