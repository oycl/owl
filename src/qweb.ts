import { ComponentData, VTree } from "./core";
import { NodeType, VDOMNode, VMultiNode, VTextNode } from "./vdom";
// import { Component } from "./component";

export interface RenderContext {
  [key: string]: any;
}

type CompiledTemplate = (this: any, tree: VTree, context: RenderContext) => void;

const templates: { [name: string]: CompiledTemplate } = {};

export function renderTemplate(tree: VTree, context: RenderContext): VTree {
  let fn = templates[tree.data.template];
  if (!fn) {
    throw new Error("qweb not implemented yet...");
  }
  return fn.call(utils, tree, context);
}

export const utils: any = {};

// -----------------------------------------------------------------------------
// demo templates
// -----------------------------------------------------------------------------
templates["<div>simple vnode</div>"] = function (tree: VTree, context: RenderContext) {
  const vn1: VDOMNode<ComponentData> = {
    type: NodeType.DOM,
    tag: "div",
    el: null,
    children: [],
    key: 2,
  };
  tree.child = vn1;
  const vn2: VTextNode = { type: NodeType.Text, text: "simple vnode", el: null };
  vn1.children.push(vn2);
};

templates["simple text node"] = function (tree: VTree, context: RenderContext) {
  const vn1: VTextNode = { type: NodeType.Text, text: "simple text node", el: null };
  tree.child = vn1;
};

templates["<div>a</div><div>b</div>"] = function (tree: VTree, context: RenderContext) {
  const vn1: VMultiNode<ComponentData> = { type: NodeType.Multi, children: [] };
  tree.child = vn1;
  const vn2: VDOMNode<ComponentData> = {
    type: NodeType.DOM,
    tag: "div",
    el: null,
    children: [],
    key: 7,
  };
  vn1.children.push(vn2);
  const vn3: VTextNode = { type: NodeType.Text, text: "a", el: null };
  vn2.children.push(vn3);
  const vn4: VDOMNode<ComponentData> = {
    type: NodeType.DOM,
    tag: "div",
    el: null,
    children: [],
    key: 9,
  };
  vn1.children.push(vn4);
  const vn5: VTextNode = { type: NodeType.Text, text: "b", el: null };
  vn4.children.push(vn5);
};

templates[`<div>Hello <t t-esc="name"/></div>`] = function (tree: VTree, context: RenderContext) {
  const vn1: VDOMNode<ComponentData> = {
    type: NodeType.DOM,
    tag: "div",
    el: null,
    children: [],
    key: 12,
  };
  tree.child = vn1;
  const vn2: VTextNode = { type: NodeType.Text, text: "Hello ", el: null };
  vn1.children.push(vn2);
  const vn3: VTextNode = { type: NodeType.Text, text: context.name, el: null };
  vn1.children.push(vn3);
};

templates[`<span><Child/></span>`] = function (tree: VTree, context: RenderContext) {
  const vn1: VDOMNode<ComponentData> = {
    type: NodeType.DOM,
    tag: "span",
    el: null,
    children: [],
    key: 15,
  };
  tree.child = vn1;
  const vn2: VTree = utils.makeComponent(tree, "Child", context);
  vn1.children.push(vn2);
};
