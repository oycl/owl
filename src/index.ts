import { Component } from "./core/component";
import { engine } from "./core/rendering_engine";
import { qweb } from "./qweb/qweb";
import { useState } from "./hooks";
import { xml } from "./tags";

const renderToString = qweb.renderToString;
const mount = engine.mount;

export { mount, renderToString, Component, xml, useState };
