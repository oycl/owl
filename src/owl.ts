import { Component } from "./component";
import { Fiber } from "./fiber";
import { RenderContext, renderTemplate } from "./qweb";
import { scheduler } from "./scheduler";
import { patch, update as updateVNode, VDataNodeMulti, VDataNode } from "./vdom";

export const enum RootType {
  Basic,
  Function,
  Component,
}

interface Data {
  fiber: Fiber;

  // todo: replace this by render function???
  template: string;
}

interface BasicRootData extends Data {
  type: RootType.Basic;
}

interface FunctionRootData extends Data {
  type: RootType.Function;
  fn: FnInstance;
}

interface ComponentRootData extends Data {
  type: RootType.Component;
  comp: Component;
}

export type RootData = BasicRootData | FunctionRootData | ComponentRootData;

export type VTree = VDataNode<RootData> | VDataNodeMulti<RootData>;

let currentType: RootType = RootType.Basic;
let currentFn: FnInstance;
let currentComp: Component;

// -----------------------------------------------------------------------------
// mount
// -----------------------------------------------------------------------------
type Fn = (env, props) => FnInstance;
type FnInstance = (props: any) => VTree;

interface MountConfig {
  target: HTMLElement;
}

export function mount(fn: Fn, config: MountConfig): Promise<VTree>;
export function mount(vnode: VTree, config: MountConfig): Promise<VTree>;
export function mount(Comp: typeof Component, config: MountConfig): Promise<VTree>;
export function mount(elem: any, config: MountConfig): Promise<VTree> {
  let vnode: VTree;

  if (typeof elem === "object") {
    vnode = elem;
  } else if (elem.prototype instanceof Component) {
    let template: string = elem.template;
    const c = new elem();
    currentType = RootType.Component;
    currentComp = c;
    vnode = render(template, c);
    vnode.hooks.create = (el) => (c.el = el);
  } else {
    const fnInstance = elem({}, {});
    currentFn = fnInstance;
    currentType = RootType.Function;
    vnode = fnInstance({});
  }
  const fiber = vnode.data.fiber;
  return scheduler.addFiber(fiber).then(() => {
    const fragment = document.createDocumentFragment();
    patch(fragment, vnode);
    config.target.appendChild(fragment);
    return vnode;
  });
}

// -----------------------------------------------------------------------------
// render
// -----------------------------------------------------------------------------

export function render(template: string, context: RenderContext = {}): VTree {
  const fiber = new Fiber(null);
  let data: RootData;
  switch (currentType) {
    case RootType.Basic:
      data = { fiber, template, type: RootType.Basic };
      break;
    case RootType.Function:
      data = { fiber, template, type: RootType.Function, fn: currentFn };
      break;
    case RootType.Component:
      data = { fiber, template, type: RootType.Component, comp: currentComp };
      break;
  }
  return renderTemplate(data, context);
}

// -----------------------------------------------------------------------------
// update
// -----------------------------------------------------------------------------

export function update(vnode: VTree, context: RenderContext = {}): Promise<void> {
  const data = vnode.data;
  const fiber = new Fiber(null);
  data.fiber = fiber;

  const newTree = renderTemplate(data, context);
  return scheduler.addFiber(fiber).then(() => {
    updateVNode(vnode, newTree);
  });
}
