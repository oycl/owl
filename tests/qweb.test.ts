import { renderToString, xml } from "../src/index";
import { compiledTemplates } from "../src/qweb/qweb";

function render(template: string, context?: any): string {
  const str = renderToString(template, context);
  expect(compiledTemplates[template].toString()).toMatchSnapshot();
  return str;
}

describe("static template", () => {
  test("simple string", () => {
    const template = xml`<t>hello vdom</t>`;
    expect(render(template)).toBe("hello vdom");
  });

  test("simple string (*)", () => {
    const template = xml`hello vdom`;
    expect(render(template)).toBe("hello vdom");
  });

  test("simple dynamic value", () => {
    const template = xml`<t><t t-esc="text"/></t>`;
    expect(render(template, { text: "hello vdom" })).toBe("hello vdom");
  });

  test("simple string, with some dynamic value", () => {
    const template = xml`<t>hello <t t-esc="text"/></t>`;
    expect(render(template, { text: "vdom" })).toBe("hello vdom");
  });

  test("no template (*)", () => {
    const template = xml``;
    expect(render(template)).toBe("");
  });

  test("empty div", () => {
    const template = xml`<div></div>`;
    expect(render(template)).toBe("<div></div>");
  });

  test("two empty divs (*)", () => {
    const template = xml`<div></div><div></div>`;
    expect(render(template)).toBe("<div></div><div></div>");
  });

});
