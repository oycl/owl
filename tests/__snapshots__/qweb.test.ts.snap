// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`static template div with a class attribute 1`] = `
"function anonymous(tree, context
) {
    // Template: <div class=\\"abc\\">word</div>
    const vn1 = tree.child = {type: 0, tag: \\"div\\", el: null, children: [], attrs: {\\"class\\":\\"abc\\"}, key: 1};
    vn1.children.push({type: 1, text: \\"word\\", el: null});
}"
`;

exports[`static template div with a class attribute with a quote 1`] = `
"function anonymous(tree, context
) {
    // Template: <div class=\\"a'bc\\">word</div>
    const vn1 = tree.child = {type: 0, tag: \\"div\\", el: null, children: [], attrs: {\\"class\\":\\"a'bc\\"}, key: 1};
    vn1.children.push({type: 1, text: \\"word\\", el: null});
}"
`;

exports[`static template div with a empty class attribute 1`] = `
"function anonymous(tree, context
) {
    // Template: <div class=\\"\\">word</div>
    const vn1 = tree.child = {type: 0, tag: \\"div\\", el: null, children: [], attrs: {}, key: 1};
    vn1.children.push({type: 1, text: \\"word\\", el: null});
}"
`;

exports[`static template div with a span child node 1`] = `
"function anonymous(tree, context
) {
    // Template: <div><span>word</span></div>
    const vn1 = tree.child = {type: 0, tag: \\"div\\", el: null, children: [], attrs: {}, key: 1};
    const vn2 = {type: 0, tag: \\"span\\", el: null, children: [], attrs: {}, key: 1};
    vn1.children.push(vn2);
    vn2.children.push({type: 1, text: \\"word\\", el: null});
}"
`;

exports[`static template div with a text node 1`] = `
"function anonymous(tree, context
) {
    // Template: <div>word</div>
    const vn1 = tree.child = {type: 0, tag: \\"div\\", el: null, children: [], attrs: {}, key: 1};
    vn1.children.push({type: 1, text: \\"word\\", el: null});
}"
`;

exports[`static template div with an arbitrary attribute with a quote 1`] = `
"function anonymous(tree, context
) {
    // Template: <div abc=\\"a'bc\\">word</div>
    const vn1 = tree.child = {type: 0, tag: \\"div\\", el: null, children: [], attrs: {\\"abc\\":\\"a'bc\\"}, key: 1};
    vn1.children.push({type: 1, text: \\"word\\", el: null});
}"
`;

exports[`static template empty div 1`] = `
"function anonymous(tree, context
) {
    // Template: <div></div>
    tree.child = {type: 0, tag: \\"div\\", el: null, children: [], attrs: {}, key: 1};
}"
`;

exports[`static template no template (*) 1`] = `
"function anonymous(tree, context
) {
    // Template: 
    tree.child = {type: 4, children:[]};
}"
`;

exports[`static template properly handle comments 1`] = `
"function anonymous(tree, context
) {
    // Template: <div>hello <!-- comment-->owl</div>
    const vn1 = tree.child = {type: 0, tag: \\"div\\", el: null, children: [], attrs: {}, key: 1};
    vn1.children.push({type: 1, text: \\"hello \\", el: null});
    vn1.children.push({type: 2, text: \\" comment\\", el: null});
    vn1.children.push({type: 1, text: \\"owl\\", el: null});
}"
`;

exports[`static template simple dynamic value 1`] = `
"function anonymous(tree, context
) {
    // Template: <t><t t-esc=\\"text\\"/></t>
    tree.child = {type: 1, text: context[\\"text\\"], el: null};
}"
`;

exports[`static template simple string (*) 1`] = `
"function anonymous(tree, context
) {
    // Template: hello vdom
    tree.child = {type: 1, text: \\"hello vdom\\", el: null};
}"
`;

exports[`static template simple string 1`] = `
"function anonymous(tree, context
) {
    // Template: <t>hello vdom</t>
    tree.child = {type: 1, text: \\"hello vdom\\", el: null};
}"
`;

exports[`static template simple string, with some dynamic value 1`] = `
"function anonymous(tree, context
) {
    // Template: <t>hello <t t-esc=\\"text\\"/></t>
    const vn1 = tree.child = {type: 4, children:[]};
    vn1.children.push({type: 1, text: \\"hello \\", el: null});
    vn1.children.push({type: 1, text: context[\\"text\\"], el: null});
}"
`;

exports[`static template two empty divs (*) 1`] = `
"function anonymous(tree, context
) {
    // Template: <div></div><div></div>
    const vn1 = tree.child = {type: 4, children:[]};
    vn1.children.push({type: 0, tag: \\"div\\", el: null, children: [], attrs: {}, key: 1});
    vn1.children.push({type: 0, tag: \\"div\\", el: null, children: [], attrs: {}, key: 1});
}"
`;
